packages/next/next-server/server/next-server.ts

```diff
@@ -485,15 +485,18 @@ export default class Server {
           type: route.type,
           name: `${route.type} ${route.source} header route`,
           fn: async (_req, res, params, _parsedUrl) => {
+            const withParams = Object.keys(params).length > 0
             for (const header of (route as Header).headers) {
               let { key, value } = header
-              if (key.includes(':')) {
-                // see `prepareDestination` util for explanation for
-                // `validate: false` being used
-                key = compilePathToRegex(key, { validate: false })(params)
-              }
-              if (value.includes(':')) {
-                value = compilePathToRegex(value, { validate: false })(params)
+              if (withParams) {
```

Comment:
We can move this check up to wrap the entire for loop since we are only looping to update the key/values and if there aren't any params then we don't have anything to update in the key/values.

Reply:
Moving this line out of the loop causes `res.setHeader` to stop working.